File: jaros/runtime/lifecycle.py
"""Agent lifecycle vocabulary for the Agent Thread Runtime (EXT-003 / REQ-1).

An agent is a lightweight in-process unit with a cheap spawn/run/teardown
lifecycle — never a service. This module defines the state vocabulary every
agent moves through and a *structural* :class:`RunnableAgent` Protocol that the
runtime depends on. The Protocol is deliberately neutral (no textual dependency
on any concrete ``agent*`` module): the pool and thread only ever rely on this
interface, so the no-server / comms checks stay satisfied.
"""

from __future__ import annotations

from enum import Enum
from typing import Protocol, runtime_checkable


# #EXT-003-REQ-1 Start
class AgentState(str, Enum):
    """The states an agent moves through during its lifecycle.

    ``spawned -> running -> (done | failed) -> torndown``. A failure is a
    terminal run outcome that is still torn down cleanly and reported.
    """

    SPAWNED = "spawned"
    RUNNING = "running"
    DONE = "done"
    FAILED = "failed"
    TORNDOWN = "torndown"


@runtime_checkable
class RunnableAgent(Protocol):
    """Structural interface the runtime depends on to host an agent.

    The runtime never imports a concrete agent class; it only requires that an
    object expose this shape. ``run()`` executes the agent body once; ``teardown()``
    releases its resources deterministically and is safe to call more than once.
    """

    id: str
    state: AgentState

    def run(self) -> None:
        """Execute the agent body exactly once."""
        ...

    def teardown(self) -> None:
        """Release all resources deterministically (idempotent)."""
        ...
# #EXT-003-REQ-1 End

File: src/runtime/lifecycle.ts
import type { Decision } from "../core/decision";

// #EXT-003-REQ-1 Start
// #EXT-003-REQ-2 Start
/**
 * Shared runtime lifecycle types for the agent thread runtime.
 *
 * These are pure runtime-plumbing contracts (state labels and the structural
 * shape of a runnable unit). They describe the runtime itself, not any agent's
 * behaviour, and carry no inter-agent channel — the pool depends on this
 * structural interface rather than reaching into the thread module directly.
 *
 * ```text
 *   spawned -> running -> (done | failed) -> torndown
 * ```
 */
export type AgentState = "spawned" | "running" | "failed" | "done" | "torndown";

/**
 * The structural contract the bounded pool drives: a lightweight in-process unit
 * with an id, an observable state, a contained `run()` that never rejects, and a
 * deterministic `teardown()`. Any implementation (e.g. an agent thread) is
 * hosted purely through this shape.
 */
export interface RunnableAgent {
  readonly id: string;
  readonly state: AgentState;
  readonly error: unknown;
  /** Run to completion; faults are contained, so this never rejects. */
  run(): Promise<Decision[]>;
  /** Release handles and join deterministically. Safe to call repeatedly. */
  teardown(): Promise<void>;
}
// #EXT-003-REQ-2 End
// #EXT-003-REQ-1 End
